state/cleanup.go

```diff
@@ -385,6 +388,46 @@ func (st *State) cleanupStorageForDyingModel(cleanupArgs []bson.Raw) (err error)
 	return nil
 }
 
+// cleanupApplication checks if all references to a dying application have been removed,
+// and if so, removes the application.
+func (st *State) cleanupApplication(applicationname string, cleanupArgs []bson.Raw) (err error) {
+	app, err := st.Application(applicationname)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Nothing to do, the application is already gone.
+			logger.Tracef("cleanupApplication(%s): application already gone", applicationname)
+			return nil
+		}
+		return errors.Trace(err)
+	}
+	if app.Life() == Alive {
+		return errors.BadRequestf("cleanupApplication requested for an application (%s) that is still alive", applicationname)
+	}
+	// We know the app is at least Dying, so check if the unit/relation counts are no longer referencing this application.
+	if app.UnitCount() > 0 || app.RelationCount() > 0 {
```

Comment:
It seems that here we are caring about relation count being > 0, but we only ever seem to add the cleanup job when removing a unit. To me this feels like we have either got our logic wrong in the expectation that relation count can be non-zero when units are zero, or that we need to also add an application cleanup job when we decrement the relation count.

Reply:
I'm pretty sure we do add a cleanup when we decrement the relation count. If you grep the diff you can see 2 calls to "newCleanupOp". One is done during RemoveUnitOperation.removeOps (which ultimately calls Application.removeUnitOps) and the other is done during Relation.removeLocalEndpointOps.